external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java

```diff
@@ -0,0 +1,654 @@
+import org.slf4j.LoggerFactory;
+
+import backtype.storm.spout.SpoutOutputCollector;
+import backtype.storm.task.TopologyContext;
+import backtype.storm.topology.OutputFieldsDeclarer;
+import backtype.storm.topology.base.BaseRichSpout;
+import backtype.storm.tuple.Fields;
+
+public class HdfsSpout extends BaseRichSpout {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HdfsSpout.class);
+
+  private Path sourceDirPath;
+  private Path archiveDirPath;
+  private Path badFilesDirPath;
+  private Path lockDirPath;
+
+  private int commitFrequencyCount = Configs.DEFAULT_COMMIT_FREQ_COUNT;
+  private int commitFrequencySec = Configs.DEFAULT_COMMIT_FREQ_SEC;
+  private int maxDuplicates = Configs.DEFAULT_MAX_DUPLICATES;
+  private int lockTimeoutSec = Configs.DEFAULT_LOCK_TIMEOUT;
+  private boolean clocksInSync = true;
+
+  private ProgressTracker tracker = new ProgressTracker();
+
+  private FileSystem hdfs;
+  private FileReader reader;
+
+  private SpoutOutputCollector collector;
+  HashMap<MessageId, List<Object> > inflight = new HashMap<>();
+  LinkedBlockingQueue<HdfsUtils.Pair<MessageId, List<Object>>> retryList = new LinkedBlockingQueue<>();
+
+  private String inprogress_suffix = ".inprogress";
+  private String ignoreSuffix = ".ignore";
+
+  private Configuration hdfsConfig;
+  private String readerType;
+
+  private Map conf = null;
+  private FileLock lock;
+  private String spoutId = null;
+
+  HdfsUtils.Pair<Path,FileLock.LogEntry> lastExpiredLock = null;
+  private long lastExpiredLockTime = 0;
+
+  private long tupleCounter = 0;
+  private boolean ackEnabled = false;
+  private int acksSinceLastCommit = 0 ;
+  private final AtomicBoolean commitTimeElapsed = new AtomicBoolean(false);
+  private final Timer commitTimer = new Timer();
+  private boolean fileReadCompletely = false;
+
+  private String configKey = Configs.DEFAULT_HDFS_CONFIG_KEY; // key for hdfs kerberos configs
+
+  public HdfsSpout() {
+  }
+
+  public Path getLockDirPath() {
+    return lockDirPath;
+  }
+
+  public SpoutOutputCollector getCollector() {
+    return collector;
+  }
+
+  public HdfsSpout withConfigKey(String configKey){
+    this.configKey = configKey;
+    return this;
+  }
+
+  public void nextTuple() {
+    LOG.debug("Next Tuple");
+    // 1) First re-emit any previously failed tuples (from retryList)
+    if (!retryList.isEmpty()) {
+      LOG.debug("Sending from retry list");
+      HdfsUtils.Pair<MessageId, List<Object>> pair = retryList.remove();
+      emitData(pair.getValue(), pair.getKey());
+      return;
+    }
+
+    if( ackEnabled  &&  tracker.size()>=maxDuplicates ) {
+      LOG.warn("Waiting for more ACKs before generating new tuples. " +
+               "Progress tracker size has reached limit {}"
+              , maxDuplicates);
+      // Don't emit anything .. allow configured spout wait strategy to kick in
+      return;
+    }
+
+    // 2) If no failed tuples, then send tuples from hdfs
+    while (true) {
```

Comment:
That loop is actually to ensure that when there is a ParseException we don't exit from the method immediately. Exiting from the method w/o emitting can trigger wait strategy. So to avoid that we retry from a new file by going back into the loop.  In other cases we such as successful parse or I/O exception, the loop will not repeat due to the return statements.